File: src/main/java/no/nav/pensjon/innsyn/poi/ss/SheetFiller.java
package no.nav.pensjon.innsyn.poi.ss;

import no.nav.pensjon.innsyn.entity.EntitySupport;
import no.nav.pensjon.innsyn.sink.CellValueSetter;
import no.nav.pensjon.innsyn.source.EntityGetter;
import org.apache.poi.ss.usermodel.*;

import java.util.function.BiConsumer;
import java.util.stream.IntStream;

/**
 * Adapted from
 * https://www.callicoder.com/java-write-excel-file-apache-poi/
 */
public class SheetFiller<T> implements SheetPopulator {

    private final static String DATE_FORMAT = "dd-MM-yyyy";
    private final String sheetName;
    private final String[] columns;
    private final EntityGetter<T> entityGetter;
    private final BiConsumer<CellValueSetter, T> createRow;

    SheetFiller(EntitySupport<T> support) {
        this.sheetName = support.getEntityDescriptor().getEntityName();
        this.columns = support.getEntityDescriptor().getPropertyNames();
        this.entityGetter = support.getEntityGetter();
        this.createRow = support.getCreateRow();
    }

    public void populateSheet(Workbook workbook) {
        Sheet sheet = workbook.createSheet(sheetName);
        createHeaderRow(sheet, createHeaderCellStyle(workbook));
        createRows(sheet, createDateCellStyle(workbook));
        fitColumnsToContentSize(sheet);
    }

    private void createRows(Sheet sheet, CellStyle dateCellStyle) {
        int rowIndex = 1;

        for (T entity : entityGetter.getEntities()) {
            createRow.accept(new RowFiller(sheet, dateCellStyle, rowIndex++), entity);
        }
    }

    private void createHeaderRow(Sheet sheet, CellStyle cellStyle) {
        Row row = sheet.createRow(0);
        createCells(row, cellStyle);
    }

    private void createCells(Row row, CellStyle cellStyle) {
        IntStream.range(0, columns.length).forEach(i -> createCell(row, i, cellStyle));
    }

    private void createCell(Row row, int index, CellStyle style) {
        Cell cell = row.createCell(index);
        cell.setCellValue(columns[index]);
        cell.setCellStyle(style);
    }

    private void fitColumnsToContentSize(Sheet sheet) {
        IntStream.range(0, columns.length).forEach(sheet::autoSizeColumn);
    }

    private static CellStyle createHeaderCellStyle(Workbook workbook) {
        CellStyle style = workbook.createCellStyle();
        style.setFont(createHeaderFont(workbook));
        return style;
    }

    private static CellStyle createDateCellStyle(Workbook workbook) {
        CellStyle style = workbook.createCellStyle();
        style.setDataFormat(workbook.getCreationHelper().createDataFormat().getFormat(DATE_FORMAT));
        return style;
    }

    private static Font createHeaderFont(Workbook workbook) {
        Font font = workbook.createFont();
        font.setBold(true);
        font.setFontHeightInPoints((short) 14);
        font.setColor(IndexedColors.RED.getIndex());
        return font;
    }
}

File: src/main/kotlin/no/nav/pensjon/innsyn/service/map/SheetFiller.kt
package no.nav.pensjon.innsyn.service.map

import no.nav.pensjon.innsyn.entity.DomainContainer
import org.apache.poi.ss.usermodel.*

/**
 * Adapted from
 * https://www.callicoder.com/java-write-excel-file-apache-poi/
 */
class SheetFiller<T> internal constructor(
        private val container: DomainContainer<T>
) {
    private val sheetName = container.entityName
    private val columns = container.propertyNames
    fun populateSheet(pid: Int, workbook: Workbook) {
        val sheet = workbook.createSheet(sheetName)
        createHeaderRow(sheet, createHeaderCellStyle(workbook))
        createRows(pid, sheet, createDateCellStyle(workbook))
        fitColumnsToContentSize(sheet)
    }

    private fun createRows(pid: Int, sheet: Sheet, dateCellStyle: CellStyle) {
        container.repository.findAllByPersonId(pid).forEachIndexed { i, e ->
            container.rowFiller(RowFiller(sheet, dateCellStyle, i + 1), e)
        }
    }

    private fun createHeaderRow(sheet: Sheet, cellStyle: CellStyle) {
        createCells(sheet.createRow(0), cellStyle)
    }

    private fun createCells(row: Row, cellStyle: CellStyle) {
        (0..columns.size).forEach { createCell(row, it, cellStyle) }
    }

    private fun createCell(row: Row, index: Int, style: CellStyle) {
        row.createCell(index).apply {
            setCellValue(columns[index])
            cellStyle = style
        }
    }

    private fun fitColumnsToContentSize(sheet: Sheet) {
        (0..columns.size).forEach(sheet::autoSizeColumn)
    }

    companion object {
        private const val DATE_FORMAT = "dd-MM-yyyy"
        private fun createHeaderCellStyle(workbook: Workbook) = workbook.createCellStyle().apply {
            setFont(createHeaderFont(workbook))
        }

        private fun createDateCellStyle(workbook: Workbook) = workbook.createCellStyle().apply {
            dataFormat = workbook.creationHelper.createDataFormat().getFormat(DATE_FORMAT)
        }

        private fun createHeaderFont(workbook: Workbook) = workbook.createFont().apply {
            bold = true
            fontHeightInPoints = 14
            color = IndexedColors.RED.getIndex()
        }
    }

}